File: Sorting/SelectionSort/selectionsort.c
#include <time.h>
#include <stdlib.h>
#include <stdio.h>

#define ARR_SIZE 10
#define BS_RAND_MIN 0
#define BS_RAND_MAX 10

/* In place sorting using a pre-malloced array */
void sort(int *array, int arraySize) {
  int min;
  int temp;

  for(int i = 0; i < arraySize - 1; i++) {
    min = i;

    for(int j = i + 1; j < arraySize; j++) {
      if(array[j] < array[min]) {
        min = j;
      }
    }

    temp = array[min];
    array[min] = array[i];
    array[i] = temp;
  }
}

int main() {
  int *array = NULL;

  array = malloc(sizeof(int) * ARR_SIZE);
  srand(time(NULL));

  printf("The original array is: [ ");
  for(int i = 0; i < ARR_SIZE; i++) {
    array[i] = rand() % (BS_RAND_MAX - BS_RAND_MIN + 1) + BS_RAND_MIN;  
    
    i == ARR_SIZE - 1 ? printf("%d ]\n", array[i]) : printf("%d, ", array[i]);
  }

  sort(array, ARR_SIZE);

  printf("The sorted array is: [ ");
  for(int i = 0; i < ARR_SIZE; i++) {
    i == ARR_SIZE - 1 ? printf("%d ]\n", array[i]) : printf("%d, ", array[i]);  
  }

  return 0;
}

File: Sorting/InPlace/SelectionSort/selectionsort.lua
rand_max = 10
rand_min = 0
array_size = 10

function sort(array)
  local min

  -- # is the length operator
  for i = 0, #array - 1 do
    min = i
    for j = i + 1, #array do
      if array[j] < array[min] then
        min = j
      end
    end
    
    temp = array[min]
    array[min] = array[i]
    array[i] = temp

  end
end

function main()
  local array = {}
  math.randomseed(os.time())

  io.write("The original array is: [ ")
  for i = 0, array_size - 1 do
    array[i] = math.random(rand_min, rand_max) % (rand_max - rand_min + 1) + rand_min
    
    if i == (array_size - 1) then
      print(string.format("%d ]", array[i]))
    else
      io.write(string.format("%d, ", array[i]))
    end
  end

  sort(array)

  io.write("The sorted array is: [ ")
  for i = 0, array_size - 1 do
    if i == (array_size - 1) then
      print(string.format("%d ]", array[i]))
    else
      io.write(string.format("%d, ", array[i]))
    end
  end
end

main()
